# scNAT

This reporsity contains code and information of data used in the paper “*scNAT: A deep learning method for integrating single cell RNA and the paired T cell receptor sequencing profiles*”. Source code for scNAT are in the [scNAT](https://github.com/biqing-zhu/scNAT/tree/main/scNAT) folder, the tutorial is in the [tutorials](https://github.com/biqing-zhu/scNAT/tree/main/tutorials) folder.

The use of scNAT to integrate scRNA-seq and scTCR-seq data allows for a more comprehensive analysis that leverages the strengths of both data types. By combining the gene expression information from scRNA-seq with the TCR repertoire diversity and clonality information from scTCR-seq, we can better understand the complex biological processes and identify novel cell subtypes that may have previously gone undetected. A figure summary is shown below.

![alt text](https://github.com/biqing-zhu/scNAT/blob/main/images/workflow.png)

### Installation

```
pip install -i https://test.pypi.org/simple/ scNAT-biqing-zhu
```
